gather-scheduler.js

```diff
@@ -19,132 +19,121 @@
-          gatherer => gatherer.afterTraceCollected(options, tracingData));
-    }).then(_ => {
-      return GatherScheduler._runPhase(gatherers,
-          gatherer => gatherer.reloadSetup(options));
-    }).then(_ => {
-      return GatherScheduler._runPhase(gatherers,
-          gatherer => gatherer.beforeReloadPageLoad(options));
-    }).then(_ => {
-      return GatherScheduler.reloadPage(driver, options);
-    }).then(_ => {
-      return GatherScheduler._runPhase(gatherers,
-          gatherer => gatherer.afterReloadPageLoad(options));
-    }).then(_ => {
-      return driver.disconnect();
-    }).then(_ => {
-      return GatherScheduler._runPhase(gatherers,
-          gatherer => gatherer.tearDown(options));
-    }).then(_ => {
+  const runPhase = phaseRunner(gatherers);
+
+  return driver.connect()
+    // Initial prep before the page load.
+    .then(_ => setupDriver(driver, gatherers, options))
+    .then(_ => runPhase(gatherer => gatherer.setup(options)))
+    .then(_ => beginPassiveCollection(driver))
+    .then(_ => runPhase(gatherer => gatherer.beforePageLoad(options)))
+
+    // Load page, gather from browser, stop profilers.
+    .then(_ => loadPage(driver, gatherers, options))
+    .then(_ => runPhase(gatherer => gatherer.afterPageLoad(options)))
+    .then(_ => endPassiveCollection(options, tracingData))
+    .then(_ => runPhase(gatherer => gatherer.afterTraceCollected(options, tracingData)))
+
+    // Reload page for SW, etc.
+    .then(_ => runPhase(gatherer => gatherer.reloadSetup(options)))
+    .then(_ => runPhase(gatherer => gatherer.beforeReloadPageLoad(options)))
+    .then(_ => reloadPage(driver, options))
+    .then(_ => runPhase(gatherer => gatherer.afterReloadPageLoad(options)))
+
+    // Finish and teardown.
+    .then(_ => driver.disconnect())
+    .then(_ => runPhase(gatherer => gatherer.tearDown(options)))
+    .then(_ => {
       // Collate all the gatherer results.
       return gatherers.map(g => g.artifact).concat(
           {networkRecords: tracingData.networkRecords},
           {traceContents: tracingData.traceContents});
     });
-  }
+}
 
-  static saveAssets(tracingData, url) {
-    const date = new Date();
-    const hostname = url.match(/^.*?\/\/(.*?)(:?\/|$)/)[1];
-    const filename = (hostname + '_' + date.toISOString() + '.trace.json')
-        .replace(/[\/\?<>\\:\*\|":]/g, '-');
-    require('fs').writeFileSync(filename, JSON.stringify(tracingData.traceContents, null, 2));
-    log('info', 'trace file saved to disk', filename);
-  }
+function saveAssets(tracingData, url) {
+  const date = new Date();
+  const hostname = url.match(/^.*?\/\/(.*?)(:?\/|$)/)[1];
+  const filename = (hostname + '_' + date.toISOString() + '.trace.json')
+      .replace(/[\/\?<>\\:\*\|":]/g, '-');
+  require('fs').writeFileSync(filename, JSON.stringify(tracingData.traceContents, null, 2));
+  log('info', 'trace file saved to disk', filename);
 }
 
-module.exports = GatherScheduler;
```

Comment:
So I prefer the static fn approach because it makes them very easy to test since they are all on GatherScheduler :)

now we either need to make a flat object and export that or export each fn :cry: